clients/src/main/java/org/apache/kafka/common/network/NetworkReceive.java

```diff
@@ -68,13 +86,32 @@ public String source() {
 
     @Override
     public boolean complete() {
-        return !size.hasRemaining() && !buffer.hasRemaining();
+        return !size.hasRemaining() && buffer != null && !buffer.hasRemaining();
     }
 
     public long readFrom(ScatteringByteChannel channel) throws IOException {
         return readFromReadableChannel(channel);
     }
 
+    @Override
+    public boolean requiredMemoryAmountKnown() {
+        return requestedBufferSize != -1;
+    }
+
+    @Override
+    public boolean memoryAllocated() {
+        return buffer != null;
+    }
+
+
+    @Override
+    public void close() throws IOException {
+        if (buffer != null && buffer != EMPTY_BUFFER) {
+            memoryPool.release(buffer);
+            buffer = null;
```

Comment:
KafkaChannel.receive is a receive _in progress_. once its compete its read out (field is nulled) and the buffer passed to a RequestChannel.Request.
so any given buffer exists either as part of an in-progress receive or as part of a completed receive (that was transferred to request channel).
given the transition happens on the same thread that would close a KafkaChannel, i dont think this would be a problem?